File: src/configuration/entities/matchmaking_settings.ts
import {
  IsBoolean,
  IsInt,
  IsString,
  IsArray,
  ValidateNested,
} from "class-validator";
import { Type } from "class-transformer";

export class Constraint {
  @IsString()
  rule: "MIN" | "MAX" | "EQUAL";

  @IsArray()
  @IsString({ each: true })
  schools: string[];

  @IsInt()
  value: number;

  @IsBoolean()
  multiple: boolean;
}

export class MatchmakingSettings {
  @IsBoolean()
  isActive: boolean;

  @IsInt()
  teamSizeMin: number;

  @IsInt()
  teamSizeMax: number;

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => Constraint)
  constraints: Constraint[];
}

File: python/matchmaking_settings.py
from typing import List
from enum import Enum
from pydantic import BaseModel

class RuleEnum(str, Enum):
    MIN = 'MIN'
    MAX = 'MAX'
    EQUAL = 'EQUAL'

class Constraint(BaseModel):
    rule: RuleEnum
    schools: List[str]
    value: int
    multiple: bool

class MatchmakingSettings(BaseModel):
    isActive: bool
    teamSizeMin: int
    teamSizeMax: int
    constraints: List[Constraint]